lib/clearance/user.rb

```diff
@@ -130,6 +132,13 @@ def normalize_email(email)
 
       private
 
+      DUMMY_PASSWORD = "*"
+
+      def prevent_timing_attack
+        new(password: DUMMY_PASSWORD)
```

Comment:
My understanding is that this will call `password=`, which will result in the password strategy running. For BCrypt that is

https://github.com/thoughtbot/clearance/blob/1dd09b0e51c839109ab4c342d0d38eeb39d27663/lib/clearance/password_strategies/bcrypt.rb#L22-L31

In the case where a user was found, we'd end up calling `authenticated?`, which would again call the password strategy

https://github.com/thoughtbot/clearance/blob/1dd09b0e51c839109ab4c342d0d38eeb39d27663/lib/clearance/password_strategies/bcrypt.rb#L16-L20

We want these two code paths to take the same amount of time to avoid leaking via timing attacks. Do `BCrypt::Password#==` and `BCrypt::Password.create` use the same process under the hood that causes them to take the same amount of time? Will this be true for all password strategies?

Reply:
For BCrypt the timings seem to be the same

```rb
require 'benchmark'
require 'benchmark/ips'
require "bcrypt"

encrypted_password = BCrypt::Password.create("password")

Benchmark.ips do |bm|
  bm.report("compare") { BCrypt::Password.new(encrypted_password) == "password" }
  bm.report("create") { BCrypt::Password.create("password") }
  bm.compare!
end
```

```
Warming up --------------------------------------
             compare     1.000  i/100ms
              create     1.000  i/100ms
Calculating -------------------------------------
             compare      4.889  (± 0.0%) i/s -     25.000  in   5.114814s
              create      4.870  (± 0.0%) i/s -     25.000  in   5.134016s

Comparison:
             compare:        4.9 i/s
              create:        4.9 i/s - 1.00x  (± 0.00) slower
```

Reply:
For argon2 as well:

```rb
require 'benchmark'
require 'benchmark/ips'
require "argon2"

encrypted_password = Argon2::Password.new.create("password")

Benchmark.ips do |bm|
  bm.report("compare") { Argon2::Password.verify_password("password", encrypted_password) }
  bm.report("create") { Argon2::Password.new.create("password") }
  bm.compare!
end
```

```Warming up --------------------------------------
             compare     1.000  i/100ms
              create     1.000  i/100ms
Calculating -------------------------------------
             compare      8.335  (± 0.0%) i/s -     42.000  in   5.044675s
              create      8.314  (± 0.0%) i/s -     42.000  in   5.057424s

Comparison:
             compare:        8.3 i/s
              create:        8.3 i/s - 1.00x  (± 0.00) slower
```

Reply:
Awesome, thanks for the benchmark! I think this is the best we can do until something changes. 👍

Reply:
I went with this approach because it was the only way I could think of fix this for any password strategy. Another way to fix would be to build a DUMMY_ENCRYPTED_PASSWORD, but the password strategy doesn't currently have an API for doing that so that would only work for our officially supported password strategies.

Reply:
Yeah, I think what you have here is excellent. If we find that some supported strategy in the future has different timings for creating and equality checks we can revisit this.